File: sdk/typescript/src/monitor.ts
import type { GraphHealth } from "./types.js";

/**
 * Real-time N★ Bit graph health monitor.
 *
 * The WebSocket endpoint sends graph health snapshots in response to any
 * message. The monitor pings every `pingInterval` ms to keep receiving updates.
 *
 * This is a monitor, not a command channel — it does not send instructions.
 */
export class Monitor {
  private ws: WebSocket | null = null;
  private pingTimer: ReturnType<typeof setInterval> | null = null;
  private stopped = false;

  constructor(
    private readonly wsUrl: string,
    private readonly onUpdate: (health: GraphHealth) => void,
    private readonly pingInterval = 5_000,
  ) {}

  start(): this {
    if (this.ws) return this;
    this._connect();
    return this;
  }

  stop(): void {
    this.stopped = true;
    if (this.pingTimer !== null) {
      clearInterval(this.pingTimer);
      this.pingTimer = null;
    }
    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }
  }

  private _connect(): void {
    // Uses native WebSocket (browser, Node.js 22+, or Deno)
    this.ws = new WebSocket(this.wsUrl);

    const ws = this.ws;
    ws.onopen = () => {
      // Send an initial ping to receive first snapshot
      ws.send("ping");
      this.pingTimer = setInterval(() => {
        if (ws.readyState === WebSocket.OPEN) {
          ws.send("ping");
        }
      }, this.pingInterval);
    };

    this.ws.onmessage = (event) => {
      try {
        const health = JSON.parse(
          typeof event.data === "string" ? event.data : String(event.data),
        ) as GraphHealth;
        this.onUpdate(health);
      } catch {
        // Unparseable frame — ignore
      }
    };

    this.ws.onclose = () => {
      if (this.pingTimer !== null) {
        clearInterval(this.pingTimer);
        this.pingTimer = null;
      }
      // Reconnect unless explicitly stopped
      if (!this.stopped) {
        setTimeout(() => this._connect(), 2_000);
      }
    };

    this.ws.onerror = () => {
      // onclose fires after onerror — reconnect logic is there
    };
  }
}

File: sdk/python/nstar_bit/monitor.py
"""
N★ Bit graph health monitor.

The WebSocket endpoint sends graph health snapshots in response to any message.
The monitor pings every `ping_interval` seconds to keep receiving updates.

This is a monitor, not a command channel.
"""
from __future__ import annotations

import asyncio
import json
from typing import Callable, Optional

import websockets
from websockets.exceptions import ConnectionClosed

from .types import GraphHealth


class Monitor:
    def __init__(
        self,
        ws_url: str,
        on_update: Callable[[GraphHealth], None],
        ping_interval: float = 5.0,
    ) -> None:
        self._ws_url = ws_url
        self._on_update = on_update
        self._ping_interval = ping_interval
        self._task: Optional[asyncio.Task] = None  # type: ignore[type-arg]
        self._stopped = False

    async def start(self) -> "Monitor":
        """Start the monitor. Connects and begins receiving health snapshots."""
        self._stopped = False
        self._task = asyncio.create_task(self._run())
        return self

    async def stop(self) -> None:
        """Stop the monitor and close the WebSocket connection."""
        self._stopped = True
        if self._task:
            self._task.cancel()
            try:
                await self._task
            except asyncio.CancelledError:
                pass
            self._task = None

    async def _run(self) -> None:
        while not self._stopped:
            try:
                async with websockets.connect(self._ws_url) as ws:
                    # Send initial ping to receive first snapshot
                    await ws.send("ping")
                    while not self._stopped:
                        try:
                            msg = await asyncio.wait_for(
                                ws.recv(), timeout=self._ping_interval + 1
                            )
                            try:
                                health = GraphHealth.from_dict(json.loads(msg))
                                self._on_update(health)
                            except (json.JSONDecodeError, KeyError):
                                pass  # Unparseable frame — ignore
                        except asyncio.TimeoutError:
                            # No message received — ping to get a fresh snapshot
                            await ws.send("ping")
            except ConnectionClosed:
                if not self._stopped:
                    await asyncio.sleep(2.0)
            except OSError:
                if not self._stopped:
                    await asyncio.sleep(2.0)
